### PR TITLE
feat(AvatarStack): add component

### PR DESCRIPTION
**Pull request by ogonkov**

### CODE REVIEW

File: src/components/ImageStack/ImageStack.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        <ul className={b(null, className)}>
+            {restItems.length > 0 ? (
+                <li key={'show-more'} className={b('item', {'overlap-size': overlapSize})}>
+                    {renderMore(restItems)}
+                </li>
+            ) : null}
+
+            {visibleItems.map((item) => (
+                <li key={item.pk} className={b('item', {'overlap-size': overlapSize})}>
+                    {renderItem(item, {itemClassName: b('item-children')})}
```
Comment (amje) [inline src/components/ImageStack/ImageStack.tsx:56]: We can clone item and patch className by ourself

**Review by amje — CHANGES_REQUESTED** [review]

File: src/components/ImageStack/ImageStack.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    return items.slice(0, getSplitIndex(items, displayCount)).reverse();
+}
+
+function getRestItems<T>(items: T[], displayCount: number) {
+    return items.slice(getSplitIndex(items, displayCount));
+}
+
+const ImageStackComponent = <T extends {pk: string}>({
+    displayCount = 2,
+    overlapSize = 's',
```
[the remark this answers is not in the document]
Comment (ogonkov) [reply src/components/ImageStack/ImageStack.tsx:26]: We have 3 predefined sizes for overlap

File: src/components/ImageStack/types.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import type React from 'react';
+
+export type OverlapSize = 's' | 'm' | 'l';
+
+export interface Props<T extends {pk: string}> {
```
[the remark this answers is not in the document]
Comment (ogonkov) [reply src/components/ImageStack/types.ts:5]: to use it as a `key` for rendering in component

File: src/components/ImageStack/ImageStack.scss [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    display: inline-flex;
+    justify-content: flex-end;
+    flex-direction: row-reverse;
+
+    margin: 0;
+    padding: 0;
+
+    &__item {
+        list-style: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg'/%3E");
```
[the remark this answers is not in the document]
Comment (amje) [reply src/components/ImageStack/ImageStack.scss:18]: Why not just `list-style-type: none;`?

File: src/components/ImageStack/types.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (amje) [reply src/components/ImageStack/types.ts:5]: That's unnecessary, item index will be enough

File: src/components/ImageStack/ImageStack.scss (hunk 4 above)
[the remark this answers is not in the document]
Comment (ogonkov) [reply src/components/ImageStack/ImageStack.scss:18]: because this will remove a11y list role

File: src/components/ImageStack/ImageStack.tsx (hunk 1 above)
Comment (amje) [reply src/components/ImageStack/ImageStack.tsx:56]: Mixing classes is bad. That could be a potential conflicts with user styles. Maybe keep a wrapper?
```jsx
<div className={b('item-children')}>{renderItem(item)}</div>
```

File: src/components/ImageStack/ImageStack.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (amje) [reply src/components/ImageStack/ImageStack.tsx:26]: It won't be enough if my image size is quite big

File: src/components/ImageStack/ImageStack.tsx (hunk 1 above)
Comment (ogonkov) [reply src/components/ImageStack/ImageStack.tsx:56]: This class should be on children

File: src/components/ImageStack/ImageStack.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (ogonkov) [reply src/components/ImageStack/ImageStack.tsx:26]: We expect that images will match sizes of our UserAvatar

File: src/components/AvatarStack/AvatarStack.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        setVisibleItems(getVisibleItems(items, displayCount));
+        setRestItems(getRestItems(items, displayCount));
+    }, [displayCount, items]);
+
+    if (!items.length) {
+        return null;
+    }
+
+    return (
+        <ul className={b(null, className)} role={'list'}>
```
Comment (amje) [inline src/components/AvatarStack/AvatarStack.tsx:47]: `ul` already has `list` role, isn't?

File: src/components/AvatarStack/__stories__/ImageStack.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,83 @@
+import React from 'react';
```
Comment (amje) [inline src/components/AvatarStack/__stories__/ImageStack.stories.tsx:1]: Filename `AvatarStack.stories.tsx`

**Review by amje — CHANGES_REQUESTED** [review]

File: src/components/AvatarStack/types.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+export interface AvatarStackProps<T extends object> {
+    items: T[];
+
+    renderItem(item: T): React.ReactNode;
+
+    renderMore(items: T[]): React.ReactNode;
+
+    /** Amount of items that should be visible */
+    displayCount?: number;
+    overlapSize?: AvatarStackOverlapSize;
```
Comment (DaffPunks) [inline src/components/AvatarStack/types.ts:14]: А точно проп должен называться `overlapSize`? Я подумал что это влияет на то, как сильно перекрывают аватарки друг друга, а оказалось что это размер всего компонента.

Может переименуем на `size`?

**Review by DaffPunks — APPROVED** [review]

File: src/components/AvatarStack/types.ts (hunk 7 above)
Comment (ogonkov) [reply src/components/AvatarStack/types.ts:14]: We didn't  set component size in any way

Comment (DaffPunks) [reply src/components/AvatarStack/types.ts:14]: Isn't it about size?
I mean use word 'overlap', but don't change overlap itself, which confuses me.

Снимок экрана 2023-08-29 в 19 06 01
Снимок экрана 2023-08-29 в 19 06 03
Снимок экрана 2023-08-29 в 19 06 06

Comment (ogonkov) [reply src/components/AvatarStack/types.ts:14]: It just storybook "feature", to demonstrate overlapSize with expected avatar size

**Review by DaffPunks — APPROVED** [review]

**Comment by amje** [comment]
I have some thoughts on this feature:
We are planning to introduce a new universal `Avatar` component that replaces current `UserAvatar` component. `Avatar` will be fully customizable, different shapes, borders, allow badges and also can be a wrapper around any ReactNode, like `Icon` can. Thus I see `AvatarStack` component as follow:

It accepts `Avatar` components as children. "More" item is an `Avatar` component with custom render and by default will be rendered depending on props `max` and `total`.  `max` prop slices passed children, `total` concats "More" item to passed children:
```jsx
<AvatarStack max={2}>
  <Avatar/>
  <Avatar/>
  <Avatar/>
  <Avatar/>
</AvatarStack>

<AvatarStack total={25}>
  <Avatar/>
  <Avatar/>
</AvatarStack>
```

Alternatively, if full control needed we can render "More" item by ourselves:

```jsx
<AvatarStack>
  <Avatar/>
  <Avatar/>
  <AvatarStack.More render={...}/>
</AvatarStack>
```

**Comment by ogonkov** [comment]
@amje i have changed component API from render props to `children` mapping. I didn't add `max` and `total` props yet, can we do it in separate PR?

File: src/components/Avatar/Avatar.scss [hunk 8]
```diff
@@ -1,9 +1,10 @@
 @use '../variables';
+@use './variables' as avatarVariables;
```
Comment (amje) [inline src/components/Avatar/Avatar.scss:2]:
```suggestion
@use './variables' as avatar-variables;
```

We prefer kebab-case in styles

File: src/components/Avatar/Avatar.scss [hunk 9]
```diff
@@ -64,24 +65,10 @@ $block: '.#{variables.$ns}avatar';
...
-            --_--size: 32px;
-        }
-
-        &_l {
-            --_--size: 42px;
-        }
-
-        &_xl {
-            --_--size: 50px;
+        @each $size_name, $size_value in avatarVariables.$sizes {
```
Comment (amje) [inline src/components/Avatar/Avatar.scss:68]:
```suggestion
        @each $size-name, $size-value in avatar-variables.$sizes {
```

File: src/components/Avatar/_variables.scss [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+$sizes: (
+    'xs': 24px,
+    's': 28px,
+    'm': 32px,
+    'l': 42px,
+    'xl': 50px,
+);
+
+$default_size: map.get($sizes, 'm');
```
Comment (amje) [inline src/components/Avatar/_variables.scss:11]:
```suggestion
$default-size: map.get($sizes, 'm');
```

File: src/components/AvatarStack/AvatarStackMoreButton.tsx [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+import type {AvatarSize} from '../Avatar';
+import {DEFAULT_AVATAR_SIZE} from '../Avatar';
+import {block} from '../utils/cn';
+
+const b = block('avatar-stack');
+
+type Props = Pick<React.HTMLProps<HTMLButtonElement>, 'className' | 'onClick' | 'aria-label'> & {
+    size?: AvatarSize;
+    count: number;
+    render?: (props: {button: React.ReactElement}) => React.ReactElement;
```
Comment (amje) [inline src/components/AvatarStack/AvatarStackMoreButton.tsx:12]: Let's use `children` for this purpose

File: src/components/AvatarStack/README.md [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+| :---------- | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :-----------: | :-----: |
+| max         | How much avatars should be visible before more button. If avatars count is only 1 short from `max`, than more button would be replaced with avatar.                   |   `number`    |    3    |
+| overlapSize | How much each item should overlap next one. `s` recommended for `UserAvatars` of sizes `xs`-`m`, `m` recomended for `l` size avatars and `l` overlap for `xl` avatars | `s`, `m`, `l` |   `s`   |
+| className   | Class name of root DOM node                                                                                                                                           |   `string`    |         |
+| children    | List of avatars, probably with some extra wrappers                                                                                                                    |  `Object[]`   |         |
```
Comment (amje) [inline src/components/AvatarStack/README.md:32]:
```suggestion
| children    | List of avatars, probably with some extra wrappers                                                                                                                    |  `React.Element[]`   |         |
```

File: src/components/AvatarStack/__stories__/AvatarStack.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    },
+    argTypes: {
+        avatarSize: {
+            control: 'select',
+            options: Object.keys(AVATAR_SIZES),
+            name: 'Size of avatar',
+            description: 'Not part of component API',
+        },
+    },
+} as Meta<ComponentType>;
```
Comment (amje) [inline src/components/AvatarStack/__stories__/AvatarStack.stories.tsx:46]: Is it necessary to create this type? `typeof AvatarStack` wouldn't work?

File: src/components/AvatarStack/__stories__/AvatarStack.stories.tsx [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+                ? React.cloneElement(child, {
+                      size: avatarSize,
+                      borderColor: 'var(--g-color-line-generic-solid)',
+                  })
+                : null,
+        ) || getChildren({avatarSize})
+    );
+}
+
+const Template: StoryFn<ComponentType> = (args) => {
```
Comment (amje) [inline src/components/AvatarStack/__stories__/AvatarStack.stories.tsx:63]: Please use newer CSF 3 format. See Button's stories as a reference

File: src/components/AvatarStack/__stories__/AvatarStack.stories.tsx [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+                    <Tooltip content={'Somehow display list of all other items'}>{button}</Tooltip>
+                )}
+            />
+        </AvatarStack>
+    );
+};
+
+export const Default = Template.bind({});
+
+export const WithOneItem = Template.bind({});
```
Comment (amje) [inline src/components/AvatarStack/__stories__/AvatarStack.stories.tsx:89]:
```suggestion
export const SingleItem = Template.bind({});
```

File: src/components/AvatarStack/__stories__/AvatarStack.stories.tsx [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+WithOneItem.args = {
+    children: getChildren({count: 1}),
+};
+
+export const WithMoreButton = Template.bind({});
```
Comment (amje) [inline src/components/AvatarStack/__stories__/AvatarStack.stories.tsx:94]:
```suggestion
export const MoreButton = Template.bind({});
```

File: src/components/AvatarStack/__stories__/AvatarStack.stories.tsx [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+    children: getChildren({count: 6}),
+};
+
+export const EdgeCase = Template.bind({});
+EdgeCase.args = {
+    children: getChildren({count: 4}),
+    max: 3,
+};
+
+export const WithButtonOverride = TemplateWithButtonOverride.bind({});
```
Comment (amje) [inline src/components/AvatarStack/__stories__/AvatarStack.stories.tsx:105]:
```suggestion
export const CustomMoreButton = TemplateWithButtonOverride.bind({});
```

File: src/components/AvatarStack/__stories__/AvatarStack.stories.tsx [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+WithOneItem.args = {
+    children: getChildren({count: 1}),
+};
+
+export const WithMoreButton = Template.bind({});
+WithMoreButton.args = {
+    children: getChildren({count: 6}),
+};
+
+export const EdgeCase = Template.bind({});
```
Comment (amje) [inline src/components/AvatarStack/__stories__/AvatarStack.stories.tsx:99]: Name isn't descriptive enough. What this story is showing?

File: src/components/AvatarStack/AvatarStackMoreButton.tsx [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+export const AvatarStackMoreButton = ({
+    className,
+    size = DEFAULT_AVATAR_SIZE,
+    onClick,
+    count,
+    'aria-label': ariaLabel,
+    render,
+}: Props) => {
+    const button = (
+        <button
```
Comment (amje) [inline src/components/AvatarStack/AvatarStackMoreButton.tsx:24]: Add focus styling for this button please

File: src/components/AvatarStack/AvatarStack.tsx [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+    const normalOverflow = moreItems >= 1;
+
+    if (!moreButton.length && hasMoreButton) {
+        const guessedSize = visibleItems[0]?.props.children.props.size;
+        moreButton.push(
+            <AvatarStackItem key="more-button">
+                <AvatarStackMoreButton
+                    size={guessedSize}
+                    count={moreItems + 1}
+                    aria-label={i18n('more')}
```
Comment (amje) [inline src/components/AvatarStack/AvatarStack.tsx:61]: Why we provide `onClick` and not use it? Default MoreButton does nothing when clicked

File: src/components/AvatarStack/__stories__/AvatarStack.stories.tsx
Comment (amje) [inline src/components/AvatarStack/__stories__/AvatarStack.stories.tsx:1]: Add visual tests for each story please

**Review by amje — CHANGES_REQUESTED** [review]

File: src/components/AvatarStack/__stories__/AvatarStack.stories.tsx (hunk 18 above)
Comment (ogonkov) [reply src/components/AvatarStack/__stories__/AvatarStack.stories.tsx:99]: Edge case when only 1 item hidden, in that case we didn't render more button, and just displays item

Comment (amje) [reply src/components/AvatarStack/__stories__/AvatarStack.stories.tsx:99]: Why we didn't render `+1` button?

File: src/components/AvatarStack/AvatarStack.tsx (hunk 20 above)
Comment (ogonkov) [reply src/components/AvatarStack/AvatarStack.tsx]: This component for user usage primary, click handler could be required

File: src/components/AvatarStack/__stories__/AvatarStack.stories.tsx (hunk 13 above)
Comment (ogonkov) [reply src/components/AvatarStack/__stories__/AvatarStack.stories.tsx:46]: Just don't want to duplicate same code in all other places it is used

File: src/components/AvatarStack/__tests__/AvatarStack.visual.test.tsx [hunk 21]
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
+
+import {test} from '~playwright/core';
+
+import {AvatarStackStories} from './stories';
+
+test.describe('AvatarStack', () => {
+    test('single avatar', async ({mount, expectScreenshot}) => {
```
Comment (amje) [inline src/components/AvatarStack/__tests__/AvatarStack.visual.test.tsx:8]:
```suggestion
    test('render story < SingleItem>', async ({mount, expectScreenshot}) => {
```

File: src/components/AvatarStack/__tests__/AvatarStack.visual.test.tsx [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+        await mount(<AvatarStackStories.SingleItem randomAvatar={false} />);
+
+        await expectScreenshot();
+    });
+
+    test('default more component', async ({mount, expectScreenshot}) => {
```
Comment (amje) [inline src/components/AvatarStack/__tests__/AvatarStack.visual.test.tsx:14]:
```suggestion
    test('render story <MoreButton>', async ({mount, expectScreenshot}) => {
```

File: src/components/AvatarStack/__tests__/AvatarStack.visual.test.tsx [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+        await mount(<AvatarStackStories.MoreButton randomAvatar={false} />);
+
+        await expectScreenshot();
+    });
+
+    test('edge case with more button omit', async ({mount, expectScreenshot}) => {
```
Comment (amje) [inline src/components/AvatarStack/__tests__/AvatarStack.visual.test.tsx:20]:
```suggestion
    test('render story <MoreButtonOmit>', async ({mount, expectScreenshot}) => {
```

**Review by amje — CHANGES_REQUESTED** [review]

File: src/components/AvatarStack/i18n/en.json [hunk 24]
```diff
@@ -0,0 +1,3 @@
+{
+  "more": ["and {{count}} more", "and {{count}} more", "and {{count}} more"]
```
[the remark this answers is not in the document]
Comment (ogonkov) [reply src/components/AvatarStack/i18n/en.json:2]: In other languages it could be different

File: src/components/AvatarStack/AvatarStackMoreButton.tsx [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+    size?: AvatarSize;
+    count: number;
+};
+
+export const AvatarStackMoreButton = React.forwardRef<
+    HTMLButtonElement,
+    AvatarStackMoreButtonProps
+>(({className, size = DEFAULT_AVATAR_SIZE, onClick, count, 'aria-label': ariaLabel}, ref) => {
+    return (
+        <button ref={ref} className={b('more-button', {size}, className)} onClick={onClick}>
```
Comment (amje) [inline src/components/AvatarStack/AvatarStackMoreButton.tsx:24]: `type="button"`

**Review by amje — CHANGES_REQUESTED** [review]

**Review by amje — APPROVED** [review]
